float32 delta_left
float32 delta_right